### ISSUE TITLE
Off-infrastructure backups that can't be deleted

### ISSUE THREAD

**Issue by grischard** [openstreetmap/operations#358]
The current data and backups could be deleted if someone malicious got their hands on a sysadmin's keys. This would wipe OSM off the map.

I would like to discuss the possibility of storing the backups with a third party where they can be written and re-read by a backup client account, but not deleted or modified. The credentials for the admin account would be stored with someone who has no access to the OpenStreetMap systems.

**Comment by tomhughes** [comment]
If your first statement was true (and I don't believe it is) then you probably shouldn't have made it in a publicly visible ticket.

The question with the second proposal are fairly obvious, among them:

* How do old backups get expired?
* Does such a service actually exist? I imagine you could construct something with S3 but it would have to use a separate account I think.
* Who is this "trusted third party" and how to we select them?

**Comment by tomhughes** [comment]
Also it really only moves the point of failure so that it is now the trusted third party whose credentials are critical.

**Comment by grischard** [comment]
Yeah, I believe a separate S3 account could do the trick, including expiration.

The trusted party could be a board member who doesn't have access to systems, i.e. not pnorman or me. These credentials would probably hardly ever be used, probably never after initial configuration.

**Comment by pnorman** [comment]
> The trusted party could be a board member who doesn't have access to systems, i.e. not pnorman or me.

I don't have access to systems. Any OWG member who isn't a sysadmin would also qualify.

**Comment by gravitystorm** [comment]
> Also it really only moves the point of failure so that it is now the trusted third party whose credentials are critical.

So the way that I would suggest it works is:

* Sysadmins do what they do now, i.e. they configure and have access to machines which can both read from the db and write to the backups on S3. Nothing changes there.
* A separate AWS account is created, which the sysadmin credentials don't give access to. This account is for the third party. If it's not the OSMF treasurer (due to having sysadmin credentials), then another OSMF board member can be that third party who controls the account.
* The third party has read-only access to the backups.
* The third party can read/write to the "cold store", which is a different location.
* The third party has no other access to the data i.e. no read/write to servers, no rights over OSMF hosting accounts (e.g. to switch off machines).
* The sysadmins have no access to the "cold store" i.e. no read or write access to the store, no management rights over the account, account managed by a non-OSMF email address etc.

This would mean that nobody has credentials that could be used to access all copies of the data + backups. It's not a bullet proof setup (happy to discuss the flaws in private), but I think it would be a useful setup.

**Comment by tomhughes** [comment]
So this second AWS account is what you're calling the "cold store" right? How does stuff move from one account to the other? It sounds like the trusted third party would have to do that so they're doing to need to run something somewhere to do that and have the technical skills to maintain it?

Is there a reason not to allow reads on the cold store, so that people can check that data really is being copied to it?

**Comment by Firefishy** [comment]
https://aws.amazon.com/about-aws/whats-new/2018/11/s3-object-lock/

**Comment by grischard** [comment]
Andy, why should sysadmin credentials not get read access to the backups? I thought this would just be our backup system, except the sysadmins wouldn't have delete/modify rights.

Reads should, I think, be allowed on the cold store so that the sysadmins with the right key can restore backups.

Sysadmin credentials could:

- Create a new backup
- Read an existing backup, e.g. for testing or restoration

Backupadmin credentials would hardly ever be used, and can:

- Configure automatic deletion and retention rules
- Configure credentials
- Delete backups manually
- Overwrite or otherwise modify a backup

**Comment by gravitystorm** [comment]
> Is there a reason not to allow reads on the cold store, so that people can check that data really is being copied to it?

Feel free to disregard that bit, I guess read access is fine.

> Sysadmin credentials could:
> 
>     * Create a new backup

I'm very wary of allowing primary credentials to write to the 'cold store'. Unless it's very carefully managed, it's easy to get into the situation where writes can occur to overwrite/truncate existing backups, rendering the cold store pointless. Something like the object lock that Grant links to could be sufficient, but I don't know enough about the edge cases to comment.

Anyway, I'm only making comments from the sidelines here, so feel free to design the system any way that suits!

**Comment by Firefishy** [comment]
Link issue: https://github.com/openstreetmap/operations/issues/746

**Comment by grischard** [comment]
Over at #864 the plan for a design was:

- Sync OSM backups to S3 on a regular automated schedule. (At creating time?)

- Cost and implement cross-site replication of backups if budgeting agreed.
